casper/src/test/scala/coop/rchain/casper/batch2/BlockValidationLogicSpec.scala

```diff
@@ -0,0 +1,150 @@
+
+      hashValid shouldBe true
+
+      val blockInValidHash = block.copy(blockHash = ByteString.copyFromUtf8("123"))
+
+      // Test invalid block hash
+      val hashInValid = BlockValidationLogic.blockHash(blockInValidHash)
+
+      hashInValid shouldBe false
+    }
+  }
+
+  "Field format validation" should "succeed on a valid block and fail on empty fields" in {
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val genesis                 = signedBlock(privateKey, publicKey)
+
+    BlockValidationLogic.formatOfFields(genesis) shouldBe true
+    BlockValidationLogic.formatOfFields(genesis.copy(blockHash = ByteString.EMPTY)) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(sig = ByteString.EMPTY)) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(sigAlgorithm = "")) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(shardId = "")) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(postStateHash = ByteString.EMPTY)) shouldBe false
+  }
+
+  // TODO: This test doesn't make sense because the signing is done by the algorithm in the ValidatorIdentity.
+  //  The algorithm specified for the block is not used
+  "Block signature validation" should "return false on unknown algorithms" in {
+    val blockUnknownAlg = getRandomBlock().copy(sigAlgorithm = "unknownAlgorithm")
+    val blockRsaAlg     = getRandomBlock().copy(sigAlgorithm = "RSA")
+
+    BlockValidationLogic.blockSignature(blockUnknownAlg) shouldBe false
+    BlockValidationLogic.blockSignature(blockRsaAlg) shouldBe false
+  }
+
+  it should "return false on invalid secp256k1 signatures" in {
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val (_, wrongPk)            = Secp256k1.newKeyPair
+    val empty                   = ByteString.EMPTY
+    val invalidKey              = "abcdef1234567890".unsafeHexToByteString
+
+    val block0 = signedBlock(privateKey, publicKey).copy(sender = empty)
+    val block1 = signedBlock(privateKey, publicKey).copy(sender = invalidKey)
+    val block2 =
+      signedBlock(privateKey, publicKey).copy(sender = ByteString.copyFrom(wrongPk.bytes))
+    val block3 = signedBlock(privateKey, publicKey).copy(sig = empty)
+    val block4 = signedBlock(privateKey, publicKey).copy(sig = invalidKey)
+    val block5 = signedBlock(privateKey, publicKey).copy(sig = block0.sig) // wrong sig
+    val blocks = Vector(block0, block1, block2, block3, block4, block5)
+
+    blocks.exists(BlockValidationLogic.blockSignature) shouldBe false
+  }
+
+  it should "return true on valid secp256k1 signatures" in {
+    val n                       = 6
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val result = (0 until n).foldLeft(true) {
+      case (res, _) =>
+        val block = signedBlock(privateKey, publicKey)
+        BlockValidationLogic.blockSignature(block) && res
+    }
+    result shouldBe true
+  }
```

Comment:
Why these two tests are not generative?

You can even create arbitrary instance for example for `(BlockMessage, Boolean)` where boolean will be expected result for signature testing. Or you can do it in two separate arbitrary instances one for correct signature and one for invalid signature.

Reply:
Yes, reasonable. It was a consequence of legacy code. In the [new approach](https://github.com/rchain/rchain/pull/3793/commits/9f5a5a84beeb6324e416d8e2ca51e0f0afcae861) I generate flags to choose which `sender` and `sig` we will use: correct or wrong from appropriate generators.